gateway/ln-gateway/src/rpc/rpc_server.rs

```diff
@@ -44,49 +39,118 @@ pub async fn run_webserver(
-            .route("/id", get(get_gateway_id));
-
-        // Authenticated, public routes used for gateway administration
-        let admin_routes = Router::new()
-            // FIXME: deprecated >= 0.3.0
-            .route("/info", post(handle_post_info))
-            .route("/info", get(info))
-            .route("/config", post(configuration))
-            .route("/balance", post(balance))
-            .route("/address", post(address))
-            .route("/withdraw", post(withdraw))
-            .route("/connect-fed", post(connect_fed))
-            .route("/leave-fed", post(leave_fed))
-            .route("/backup", post(backup))
-            .route("/restore", post(restore))
-            .route("/set_configuration", post(set_configuration))
-            .layer(ValidateRequestHeaderLayer::bearer(&gateway_config.password));
-        (public_routes, admin_routes)
-    } else {
-        let public_routes = Router::new()
-            .route("/set_configuration", post(set_configuration))
-            .route("/config", get(configuration))
-            // FIXME: deprecated >= 0.3.0
-            .route("/info", post(handle_post_info))
-            .route("/info", get(info));
-        let admin_routes = Router::new();
-        (public_routes, admin_routes)
-    };
+/// Extracts the Bearer token from the Authorization header of the request.
+fn extract_bearer_token(request: &Request) -> Result<String, StatusCode> {
+    let headers = request.headers();
+    let auth_header = headers.get(header::AUTHORIZATION);
+    if let Some(header_value) = auth_header {
+        let auth_str = header_value
+            .to_str()
+            .map_err(|_| StatusCode::UNAUTHORIZED)?;
+        let token = auth_str.trim_start_matches("Bearer ").to_string();
+        return Ok(token);
+    }
+
+    Err(StatusCode::UNAUTHORIZED)
+}
+
+/// Middleware to authenticate an incoming request. Routes that are
+/// authenticated with this middleware always require a Bearer token to be
+/// supplied in the Authorization header.
+async fn admin_authentication_middleware(
+    Extension(gateway): Extension<Gateway>,
+    request: Request,
+    next: Next,
+) -> Result<impl IntoResponse, StatusCode> {
+    let gateway_config = gateway.gateway_config.read().await.clone();
+    // These routes are not available unless the gateway's configuration is set.
+    let gateway_password = gateway_config.ok_or(StatusCode::NOT_FOUND)?.password;
+    let token = extract_bearer_token(&request)?;
+    if gateway_password == token {
+        return Ok(next.run(request).await);
+    }
+
+    Err(StatusCode::UNAUTHORIZED)
+}
+
+/// Middleware to authenticate an incoming request. Routes that are
+/// authenticated with this middleware are un-authenticated if the gateway has
+/// not yet been configured. After the gateway is configured, this middleware
+/// enforces that a Bearer token must be supplied in the Authorization header.
+async fn admin_after_config_authentication_middleware(
```

Comment:
name suggestion: `auth_after_config_middleware_fn`

Reply:
Made this change but dropped the `_fn` part, should be obvious that it is a function :)